src/DraxList.tsx

```diff
@@ -438,48 +442,73 @@ export const DraxList = <T extends unknown>(
 		[horizontal, itemCount, originalIndexes],
 	);
 
-	// Stop scrolling, and potentially update shifts and reorder data.
+	// Stop auto-scrolling, and potentially update shifts and reorder data.
 	const handleInternalDragEnd = useCallback(
 		(
-			dragged?: DraxEventViewData,
-			receiver?: DraxEventViewData,
+			eventData: DraxMonitorEventData | DraxMonitorEndEventData | DraxMonitorDragDropEventData,
+			totalDragEnd: boolean,
 		): DraxProtocolDragEndResponse => {
 			// Always stop auto-scroll on drag end.
 			scrollStateRef.current = AutoScrollDirection.None;
 			stopScroll();
 
-			// If the list is reorderable, handle shifts and reordering.
-			if (reorderable) {
+			const { dragged, receiver } = eventData;
+
+			// Check if we need to handle this drag end.
+			if (reorderable && dragged.parentId === id) {
 				// Determine list indexes of dragged/received items, if any.
-				const fromPayload = dragged && (dragged.parentId === id)
-					? (dragged.payload as ListItemPayload)
-					: undefined;
-				const toPayload = (fromPayload !== undefined && receiver && receiver.parentId === id)
+				const fromPayload = dragged.payload as ListItemPayload;
+				const toPayload = receiver?.parentId === id
 					? (receiver.payload as ListItemPayload)
 					: undefined;
 
-				if (fromPayload !== undefined) {
-					// If dragged item was ours, reset shifts.
-					resetShifts();
-					if (toPayload !== undefined) {
-						// If dragged item and received item were ours, reorder data.
-						// console.log(`moving ${fromPayload.index} -> ${toPayload.index}`);
-						const snapbackTarget = calculateSnapbackTarget(fromPayload, toPayload);
-						const { index: fromIndex, originalIndex: fromOriginalIndex } = fromPayload;
-						const { index: toIndex, originalIndex: toOriginalIndex } = toPayload;
-						if (data) {
-							const newOriginalIndexes = originalIndexes.slice();
-							newOriginalIndexes.splice(toIndex, 0, newOriginalIndexes.splice(fromIndex, 1)[0]);
-							setOriginalIndexes(newOriginalIndexes);
-							onItemReorder?.({
-								fromIndex,
-								toIndex,
-								fromItem: data[fromOriginalIndex],
-								toItem: data[toOriginalIndex],
-							});
-						}
-						return snapbackTarget;
+				const { index: fromIndex, originalIndex: fromOriginalIndex } = fromPayload;
+				const { index: toIndex, originalIndex: toOriginalIndex } = toPayload ?? {};
+				const toItem = toOriginalIndex ? data?.[toOriginalIndex] : undefined;
+
+				// Reset all shifts and call callback, regardless of whether toPayload exists.
+				resetShifts();
+				if (totalDragEnd) {
```

Comment:
What does `totalDragEnd` signify? I see there is a case for it being `totalDragEnd` and one where there isn't a `totalDragEnd` that looks like the drag moved positions. Is that right?

Reply:
For reference, the entire list is wrapped in a Drax monitor. An internal drag end means we are no longer dragging a list item from position to position, which can be triggered by three different monitor events:

`onMonitorDragExit`: the drag is still continuing but has exited the list zone. This is not a "total drag end".
`onMonitorDragEnd`: the total drag has ended within the list zone, not over a receiver.
`onMonitorDragDrop`: the total drag has ended within the list zone, over a receiver (a new position).

Since there is a lot of common logic for these cases, it's in the `handleInternalDragEnd` function. We use the `totalDragEnd` boolean to distinguish whether the drag gesture has actually ended and adjust our behavior accordingly:

- If it is a total drag end, we trigger the `onItemDragEnd` callback to signify that the list item drag has ended.

- If it is not a total drag end, we trigger the `onItemDragPositionChange` callback to signify that the list item has been dragged to an undefined position.

The naming `totalDragEnd` was somewhat arbitrary. I essentially just wanted to distinguish that certain subset of events. Looking at it again now with fresh eyes, it might have made more sense to call it `dragGestureEnd`. I might change it, though the name is only used internally here.